File: c/caesar.c
#include <stdlib.h>
#include <string.h>
#include <assert.h>


char* encrypt(const char* text, const size_t k, const size_t n)
{
    const size_t tlength = strlen(text);
    char* result = (char*)malloc(tlength);

    for (register int i = 0; i < tlength; ++i)
    {
        result[i] = (text[i] + k) % n;
    }

    return result;
}


char* decrypt(const char* text, const size_t k, const size_t n)
{
    const size_t tlength = strlen(text);
    char* result = (char*)malloc(tlength);

    for (register int i = 0; i < tlength; ++i)
    {
        result[i] = (text[i] + n - k) % n;
    }

    return result;
}


int main(int argc, char* argv[])
{
    assert(strcmp(encrypt("cryptography", 3, 256), "fu|swrjudsk|") == 0);
    assert(strcmp(decrypt("fu|swrjudsk|", 3, 256), "cryptography") == 0);
    return 0;
}

File: python/caesar.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
"""
caesar.py: Caesar algorithm.
Formulas:
    1) Ek(i) = (i+k) mod n.
    2) Dk(i) = (i+n-k) mod n.
"""


def encrypt(text: str, k: int = 3, n: int = 256) -> str:
    """
    encrypt: Caesar encryption algorithm for ASCII codes.
    """
    return ''.join([chr((ord(x) + k) % n) for x in text])


def decrypt(text: str, k: int = 3, n: int = 256) -> str:
    """
    decrypt: Caesar decryption algorithm for ASCII codes.
    """
    return ''.join([chr((ord(x) + n - k) % n) for x in text])


if __name__ == '__main__':
    assert encrypt('cryptography') == 'fu|swrjudsk|'
    assert decrypt('fu|swrjudsk|') == 'cryptography'
